metaseq/tasks/streaming_language_modeling.py

```diff
@@ -605,32 +620,42 @@ def load_dataset(self, split: str, epoch=1, combine=False, **kwargs):
-                eod=self.cm3_break_ind,
-                dataset=dataset,
-                # We generate blocks with one extra token, so that we have a target
-                # for the final input token. This results in slight data loss.
-                block_size=self.args.tokens_per_sample + 1,
-                break_mode=self.args.sample_break_mode,
-                # we drop the remainder block during training
-                drop_last=(split == "train"),
-                padding_idx=self.source_dictionary.pad(),
-                seed=self.args.seed,
-                percent_full_document_rotation=self.args.cm3_percent_full_document_rotation,
+            self.datasets[split] = DatasetWithShardInformation(
+                CausalMaskedDocumentToSequenceDataset(
+                    sentinel_token_expectation=self.args.cm3_lambda_sentinel_tokens,
+                    sentinel_tokens=self.cm3_sentinel_tokens_ind,
+                    sentinel_method=self.cm3_sentinel_type,
+                    sentinel_eos=self.cm3_sentinel_end_ind,
+                    allow_rotation_across_eod=self.args.cm3_allow_across_eod_boundaries,
+                    eod=self.cm3_break_ind,
+                    dataset=dataset,
+                    # We generate blocks with one extra token, so that we have a target
+                    # for the final input token. This results in slight data loss.
+                    block_size=self.args.tokens_per_sample + 1,
+                    break_mode=self.args.sample_break_mode,
+                    # we drop the remainder block during training
+                    drop_last=(split == "train"),
+                    padding_idx=self.source_dictionary.pad(),
+                    seed=self.args.seed,
+                    percent_full_document_rotation=self.args.cm3_percent_full_document_rotation,
+                ),
+                is_sharded=is_sharded,
+                shard_id=shard_id,
+                num_shards=num_shards,
             )
         else:
-            self.datasets[split] = DocumentToSequenceDataset(
-                dataset,
-                block_size=self.args.tokens_per_sample + 1,
-                break_mode=self.args.sample_break_mode,
-                drop_last=(split == "train"),
-                padding_idx=self.source_dictionary.pad(),
-                seed=self.args.seed,
+            self.datasets[split] = DatasetWithShardInformation(
+                DocumentToSequenceDataset(
+                    dataset,
+                    block_size=self.args.tokens_per_sample + 1,
+                    break_mode=self.args.sample_break_mode,
+                    drop_last=(split == "train"),
+                    padding_idx=self.source_dictionary.pad(),
+                    seed=self.args.seed,
+                ),
+                is_sharded=is_sharded,
+                shard_id=shard_id,
+                num_shards=num_shards,
```

Comment:
Move to outside if-else to remove code duplication?

self.datasets[split] = DatasetWithShardInformation(self.datasets[split] , is_sharded ....)

Reply:
sg